src/xrpld/ledger/detail/View.cpp

```diff
@@ -1831,39 +2041,155 @@ requireAuth(ReadView const& view, Issue const& issue, AccountID const& account)
+    {
+        // Either account is issuer and no authorization is needed, or it has
+        // credentials maching DomainID stored in MPT issuance.
+        // Note: the credentials on which basis we return tesSUCCESS might have
+        // expired; the user must call verifyAuth to check with ApplyView.
+        return tesSUCCESS;
+    }
+
+    auto sleToken = maybeToken->token;
+    auto sleIssuance = maybeToken->issuance;
 
     // mptoken must be authorized if issuance enabled requireAuth
-    if (sleIssuance->getFieldU32(sfFlags) & lsfMPTRequireAuth &&
+    if ((sleIssuance->getFieldU32(sfFlags) & lsfMPTRequireAuth) &&
         !(sleToken->getFlags() & lsfMPTAuthorized))
         return tecNO_AUTH;
 
     return tesSUCCESS;
 }
 
+[[nodiscard]] TER
+verifyAuth(
+    ApplyView& view,
+    MPTIssue const& mptIssue,
+    AccountID const& account,
+    XRPAmount const& priorBalance,  // for MPToken authorization
+    beast::Journal j)
+{
+    auto const mptIssuanceID = mptIssue.getMptID();
+    auto const sleIssuance = view.read(keylet::mptIssuance(mptIssuanceID));
+    if (!sleIssuance)
+        return tefINTERNAL;  // Should have called requireAuth earlier
+
+    if (account == sleIssuance->at(sfIssuer))
+        return tesSUCCESS;  // Won't create MPToken for the token issuer
+
+    auto const sleToken = view.read(keylet::mptoken(mptIssuanceID, account));
+    bool const domainCheck =
+        (sleToken && sleToken->getFlags() & lsfMPTDomainCheck);
+
+    bool authorizedByDomain = false;
+    if (domainCheck || sleToken == nullptr)
+    {
+        // We check DomainID if:
+        // 1. Token not found or
+        // 2. Token found and has lsfMPTDomainCheck flag
+        auto const maybeDomainID = sleIssuance->at(~sfDomainID);
+        authorizedByDomain = maybeDomainID.has_value() &&
+            verifyDomain(view, account, *maybeDomainID, j) == tesSUCCESS;
+    }
+
+    if (authorizedByDomain && sleToken == nullptr)  // && !domainCheck, implied
+    {
+        // Create MPToken with the lsfMPTDomainCheck flag set
+        if (auto const err = MPTokenAuthorize::authorize(
+                view,
+                j,
+                {
+                    .priorBalance = priorBalance,
+                    .mptIssuanceID = mptIssuanceID,
+                    .accountID = account,
+                    .flags = 0,
+                });
+            !isTesSuccess(err))
+            return err;
+
+        auto sleMpt = view.peek(keylet::mptoken(mptIssuanceID, account));
+        XRPL_ASSERT(sleMpt, "ripple::verifyAuth : found new MPToken");
+        std::uint32_t const flags = sleMpt->getFieldU32(sfFlags);
+        sleMpt->setFieldU32(sfFlags, flags | lsfMPTDomainCheck);
+        view.update(sleMpt);
+
+        return tesSUCCESS;
+    }
+    else if (!authorizedByDomain && domainCheck)  // && sleToken, implied
+    {
+        // We will only delete MPToken here if it has 0 balance
+        [[maybe_unused]] auto _ = MPTokenAuthorize::authorize(
```

Comment:
Do we need to delete it? Later if permissions change then we'd have to create it again. Is there any harm in keeping `MPToken` if it has 0 balance?

Reply:
I think we do - otherwise it won't be ever removed by anyone else. Unless MPTokens are all deleted when MPTokenIssuance is deleted ?

Reply:
No, we don't delete all MPToken when MPTokenIssuance is deleted. This could be computationally unbound. The argument in this case is that it's up to the holders to delete MPTToken. In case of SAV, MPToken is automatically created. So yes, it makes sense automatically delete even though there is some inefficiency.